compiler/src/iree/compiler/InputConversion/Common/IREECheckIllegalDialects.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-dialects/Dialect/Input/InputDialect.h"
+#include "iree-dialects/Dialect/Input/InputOps.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowDialect.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/HAL/IR/HALTypes.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilTypes.h"
+#include "iree/compiler/InputConversion/Common/PassDetail.h"
+#include "iree/compiler/InputConversion/Common/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir::iree_compiler {
+
+namespace {
+
+struct IREECheckIllegalDialectsPass
+    : public IREECheckIllegalDialectsBase<IREECheckIllegalDialectsPass> {
+  void runOnOperation() override {
+    auto operation = getOperation();
+    llvm::DenseSet<Operation *> errors;
+    llvm::DenseSet<StringRef> invalid{"tf", "torch", "stablehlo", "mhlo",
+                                      "chlo"};
```

Comment:
How about extending these passes instead?

* https://github.com/openxla/iree/blob/main/compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/VerifyCompilerInputLegality.cpp
* https://github.com/openxla/iree/blob/main/compiler/plugins/input/TOSA/tosa-iree/InputConversion/VerifyCompilerTOSAInputLegality.cpp
* (and add one for Torch)

We may actually want to remove or refactor https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Dialect/Flow/Transforms/VerifyInputLegality.cpp at this point too.

Each input plugin that provides a conversion from some other dialect(s) knows best which ops might exist and _should be converted_. After that point, the core project doesn't have as much information about program structure, user intent, possible dialects, etc.

The other failure mode is when an input plugin is not used, causing ops to not be converted, but I'd rather fix that directly (with a good error message or better auto selection during input conversion) vs add a patch further down the pipeline with a less specific error check.

As Ben says, custom dialects also factor in, and I'd want to be careful about disallowing things like user-authored HAL executables with 'spirv' ops in them (e.g. if we only allows 'flow'/'hal'/'stream'/'input'/etc.).

Reply:
yeah, we already have people passing down cudnn dialects and such

Reply:
I'll put this work in my queue, unless Rob picks it up again.

Plan is roughly as I described in this comment.

Reply:
Ah, I found a catch here - if an input program is _only_ TF/MHLO with no StableHLO and the "auto" input type is used, the StableHLO input plugin will be bypassed entirely, so `compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/VerifyCompilerInputLegality.cpp` won't even have a chance to check...

In that case a generic pass like in this PR could actually catch things.

Now that would be less of an issue if we just ran all input conversions regardless of what the module contents were, but I don't think we want to do that...

Reply:
At least in the case of https://github.com/openxla/iree/issues/16021, the module contained both stablehlo and tf ops:
![image](https://github.com/openxla/iree/assets/4010439/5b52d147-68cd-446e-a6dd-ed19833ad186)